stan/math/rev/mat/fun/multiply.hpp

```diff
@@ -1,23 +1,407 @@
 #ifndef STAN_MATH_REV_MAT_FUN_MULTIPLY_HPP
 #define STAN_MATH_REV_MAT_FUN_MULTIPLY_HPP
 
-#include <stan/math/rev/core.hpp>
-#include <stan/math/rev/mat/fun/Eigen_NumTraits.hpp>
-#include <stan/math/rev/mat/fun/typedefs.hpp>
+#include <boost/utility/enable_if.hpp>
+#include <boost/type_traits.hpp>
 #include <stan/math/prim/mat/fun/Eigen.hpp>
 #include <stan/math/prim/mat/fun/typedefs.hpp>
-#include <stan/math/prim/mat/err/check_multiplicable.hpp>
+#include <stan/math/rev/scal/fun/value_of_rec.hpp>
+#include <stan/math/rev/scal/fun/value_of.hpp>
 #include <stan/math/rev/mat/fun/to_var.hpp>
-#include <stan/math/rev/mat/fun/dot_product.hpp>
+#include <stan/math/rev/core.hpp>
+#include <stan/math/prim/mat/fun/value_of_rec.hpp>
+#include <stan/math/prim/mat/err/check_multiplicable.hpp>
+#include <stan/math/prim/scal/err/check_not_nan.hpp>
 #include <boost/utility/enable_if.hpp>
 #include <boost/type_traits.hpp>
 #include <boost/math/tools/promotion.hpp>
 
 namespace stan {
   namespace math {
 
+    template <typename TA, int RA_, int CA_, typename TB, int CB_>
+    class multiply_mat_vari : public vari {
+    public:
+      int RA, CA, CB, Asize, Bsize;  // A.cols() = B.rows()
+      double* Ad_a;
+      double* Bd_a;
+      vari** variRefA_;
+      vari** variRefB_;
+      vari** variRefAB_;
+
+      /* ctor for multiply function
+       *
+       * Stores pointers to varis for A
+       * Stores pointers to varis for B
+       * Stores pointers to varis for AB
+       * All memory allocated on ChainableStack
+       *
+       * @tparam TA Scalar type for matrix A
+       * @tparam RA_ Rows for matrix A
+       * @tparam CA_ Columns for matrix A, Rows for matrix B
+       * @tparam TB Scalar type for matrix B
+       * @tparam CB_ Columns for matrix B
+       * @param matrix A
+       * @param matrix B
+       * */
+     multiply_mat_vari(const Eigen::Matrix<TA, RA_, CA_>& A,
+                       const Eigen::Matrix<TB, CA_, CB_>& B)
+        : vari(0.0),
+          RA(A.rows()), CA(A.cols()),
+          CB(B.cols()), Asize(A.size()), Bsize(B.size()),
```

Comment:
Right --- no trailing underscore for template params.

> On Aug 11, 2016, at 4:09 AM, Rob Trangucci <notifications@github.com> wrote:
> 
> In stan/math/rev/mat/fun/multiply.hpp:
> 
> > +       * Stores pointers to varis for AB
> > +       * All memory allocated on ChainableStack
> > +       *
> > +       * @tparam TA Scalar type for matrix A
> > +       * @tparam RA_ Rows for matrix A
> > +       * @tparam CA_ Columns for matrix A, Rows for matrix B
> > +       * @tparam TB Scalar type for matrix B
> > +       * @tparam CB_ Columns for matrix B
> > +       * @param matrix A
> > +       * @param matrix B
> > +       * */
> > +     multiply_mat_vari(const Eigen::Matrix<TA, RA_, CA_>& A,
> > +                       const Eigen::Matrix<TB, CA_, CB_>& B)
> > +        : vari(0.0),
> > +          RA(A.rows()), CA(A.cols()),
> > +          CB(B.cols()), Asize(A.size()), Bsize(B.size()),
> 
> Agreed; that'll allow me to use RA, CA, CB for template parameters that agree with the template parameters for multiply(A, B) down below.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>